sklearn/utils/validation.py

```diff
@@ -565,10 +584,33 @@ def check_array(array, accept_sparse=False, dtype="numeric", order=None,
     return array
 
 
-def check_X_y(X, y, accept_sparse=False, dtype="numeric", order=None,
-              copy=False, force_all_finite=True, ensure_2d=True,
-              allow_nd=False, multi_output=False, ensure_min_samples=1,
-              ensure_min_features=1, y_numeric=False,
+def _check_large_sparse(X, accept_large_sparse=False):
+    """Raise a TypeError if X has 64bit indices and accept_large_sparse=False
+    """
+    if not (accept_large_sparse and LARGE_SPARSE_SUPPORTED):
+        supported_indices = ["int32"]
+        if X.getformat() == "coo":
+            index_keys = ['col', 'row']
+        elif X.getformat() in ["csr", "csc", "bsr"]:
+            index_keys = ['indices', 'indptr']
+        else:
+            return
+        for key in index_keys:
+            indices_datatype = getattr(X, key).dtype
+            if (indices_datatype not in supported_indices):
+                if not LARGE_SPARSE_SUPPORTED:
+                    raise TypeError("Scipy version %s does not support large"
+                                    " indices, please upgrade your scipy"
+                                    " to 0.14.0 or above" % scipy_version)
+                raise TypeError("Only sparse matrices with 32-bit integer"
+                                " indices are accepted. Got %s indices."
+                                % indices_datatype)
```

Comment:
Other scikit-learn input validation checks tend to raise `ValueError` when the type of the container is ok (array, sparse matrix, dataframe) but the dtype is invalid:

```python
>>> import numpy as np
>>> from sklearn.linear_model import LogisticRegression
>>> LogisticRegression().fit(np.array([['invalid'], ['invalid']], dtype=object), [0, 1])
Traceback (most recent call last):
  File "<ipython-input-8-d601c7dabfdc>", line 1, in <module>
    LogisticRegression().fit(np.array([['invalid'], ['invalid']], dtype=object), [0, 1])
  File "/home/ogrisel/code/scikit-learn/sklearn/linear_model/logistic.py", line 1218, in fit
    order="C")
  File "/home/ogrisel/code/scikit-learn/sklearn/utils/validation.py", line 671, in check_X_y
    ensure_min_features, warn_on_dtype, estimator)
  File "/home/ogrisel/code/scikit-learn/sklearn/utils/validation.py", line 494, in check_array
    array = np.asarray(array, dtype=dtype, order=order)
  File "/home/ogrisel/.virtualenvs/py36/lib/python3.6/site-packages/numpy/core/numeric.py", line 492, in asarray
    return array(a, dtype, copy=False, order=order)
ValueError: could not convert string to float: 'invalid'
```

Therefore we might want to raise `ValueError` here.

Reply:
Hmm. We raise TypeError for wrong sparse format. I can change to ValueError...?

Reply:
I can change this, but I feel like TypeError is more precise.